File: android/app/src/main/java/com/expensetracker/data/DemoData.kt
package com.expensetracker.data

import java.text.NumberFormat
import java.text.SimpleDateFormat
import java.util.*

// Data Models
data class Transaction(
    val id: String = UUID.randomUUID().toString(),
    val title: String,
    val amount: Double,
    val isExpense: Boolean,
    val category: String,
    val date: Date,
    val familyMember: String
)

data class Category(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val icon: String,
    val color: String,
    val isExpense: Boolean
)

data class SavingsGoal(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val currentAmount: Double,
    val targetAmount: Double
) {
    val progress: Float get() = (currentAmount / targetAmount).toFloat()
    val progressPercent: String get() = "${(progress * 100).toInt()}%"
}

data class FamilyMember(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val role: String,
    val avatar: String
)

data class BudgetCategory(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val limit: Double,
    val spent: Double
) {
    val progress: Float get() = (spent / limit).toFloat()
}

data class RecurringExpense(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val amount: Double,
    val category: String,
    val frequency: String,
    val nextDate: Date
)

// New data models matching backend DTOs
data class LedgerCreateDTO(
    val type: String, // "EXPENSE" or "INCOME"
    val amountMinor: Long,
    val currency: String,
    val categoryId: String,
    val memberId: String,
    val occurredAt: String, // ISO 8601 timestamp
    val notes: String? = null
)

data class RecurringRuleDTO(
    val id: String? = null,
    val familyId: String,
    val name: String,
    val type: String, // "EXPENSE" or "INCOME"
    val amountMinor: Long,
    val currency: String,
    val categoryId: String,
    val memberId: String,
    val frequency: FrequencyDTO,
    val startDate: String, // ISO date
    val endDate: String? = null,
    val timezone: String,
    val nextRunAt: String? = null,
    val isPaused: Boolean = false
)

data class FrequencyDTO(
    val unit: String, // "WEEKLY", "MONTHLY", "YEARLY"
    val interval: Int = 1,
    val byMonthDay: List<Int>? = null
)

data class BudgetDTO(
    val id: String? = null,
    val name: String,
    val period: PeriodDTO,
    val overallLimitMinor: Long,
    val includeRecurring: Boolean = true,
    val alertThresholdPct: Int = 80,
    val perCategory: List<CategoryBudgetDTO> = emptyList()
)

data class PeriodDTO(
    val type: String, // "MONTH", "QUARTER", "YEAR", "CUSTOM"
    val start: String, // ISO date
    val end: String // ISO date
)

data class CategoryBudgetDTO(
    val categoryId: String,
    val limitMinor: Long
)

data class BudgetSpendDTO(
    val period: PeriodDTO,
    val overall: SpendingSummaryDTO,
    val byCategory: List<CategorySpendingDTO>
)

data class SpendingSummaryDTO(
    val limitMinor: Long,
    val spentMinor: Long
)

data class CategorySpendingDTO(
    val categoryId: String,
    val limitMinor: Long,
    val spentMinor: Long
)

data class ReportSummaryDTO(
    val totalIncomeMinor: Long,
    val totalExpensesMinor: Long,
    val balanceMinor: Long,
    val perCategory: List<CategorySummaryDTO>,
    val perMonth: List<MonthlySummaryDTO>,
    val recentTransactions: List<TransactionSummaryDTO>
)

data class CategorySummaryDTO(
    val categoryId: String,
    val spentMinor: Long
)

data class MonthlySummaryDTO(
    val month: String, // "2025-10"
    val incomeMinor: Long,
    val expenseMinor: Long
)

data class TransactionSummaryDTO(
    val id: String,
    val categoryId: String,
    val amountMinor: Long
)

data class ExportResponse(
    val exportId: String
)

// Demo Data Object
object DemoData {
    const val totalIncome: Double = 12500.00
    const val totalExpenses: Double = 8200.00
    
    val balance: Double get() = totalIncome - totalExpenses
    
    val transactions = listOf(
        Transaction(
            title = "Salary",
            amount = 5000.00,
            isExpense = false,
            category = "Income",
            date = Date(),
            familyMember = "John"
        ),
        Transaction(
            title = "Grocery Shopping",
            amount = 156.78,
            isExpense = true,
            category = "Groceries",
            date = Date(),
            familyMember = "Sarah"
        ),
        Transaction(
            title = "Gas Station",
            amount = 45.50,
            isExpense = true,
            category = "Transportation",
            date = Date(),
            familyMember = "John"
        ),
        Transaction(
            title = "Restaurant",
            amount = 89.32,
            isExpense = true,
            category = "Dining",
            date = Date(),
            familyMember = "Sarah"
        ),
        Transaction(
            title = "Freelance Work",
            amount = 800.00,
            isExpense = false,
            category = "Income",
            date = Date(),
            familyMember = "John"
        )
    )
    
    val categories = listOf(
        Category("Groceries", "cart", "accentSuccess", true),
        Category("Transportation", "car", "accentWarning", true),
        Category("Dining", "fork.knife", "accentInfo", true),
        Category("Entertainment", "tv", "primary", true),
        Category("Utilities", "house", "expense", true),
        Category("Salary", "dollarsign.circle", "income", false),
        Category("Freelance", "briefcase", "income", false)
    )
    
    val savingsGoals = listOf(
        SavingsGoal("Emergency Fund", 2500.00, 5000.00),
        SavingsGoal("Vacation", 1200.00, 3000.00),
        SavingsGoal("New Car", 4500.00, 6000.00),
        SavingsGoal("Home Renovation", 900.00, 3000.00)
    )
    
    val familyMembers = listOf(
        FamilyMember("John Smith", "Admin", "person.circle.fill"),
        FamilyMember("Sarah Smith", "Member", "person.circle.fill"),
        FamilyMember("Emma Smith", "Member", "person.circle.fill")
    )
    
    val budgetCategories = listOf(
        BudgetCategory("Groceries", 500.00, 325.00),
        BudgetCategory("Transportation", 300.00, 245.00),
        BudgetCategory("Dining", 200.00, 89.32),
        BudgetCategory("Entertainment", 150.00, 67.50)
    )
    
    val recurringExpenses = listOf(
        RecurringExpense(
            "Netflix",
            15.99,
            "Entertainment",
            "Monthly",
            Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, 5) }.time
        ),
        RecurringExpense(
            "Gym Membership",
            29.99,
            "Health",
            "Monthly",
            Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, 12) }.time
        ),
        RecurringExpense(
            "Phone Bill",
            89.00,
            "Utilities",
            "Monthly",
            Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, 8) }.time
        )
    )
}

// Extension functions for formatting
fun Double.formatAsCurrency(): String {
    val formatter = NumberFormat.getCurrencyInstance(Locale.US)
    return formatter.format(this)
}

fun Double.formatAsPercent(): String {
    return "${(this * 100).toInt()}%"
}

fun Date.formatAsMonthDay(): String {
    val formatter = SimpleDateFormat("MM/dd", Locale.US)
    return formatter.format(this)
}

fun Date.formatAsShort(): String {
    val formatter = SimpleDateFormat("M/d/yyyy", Locale.US)
    return formatter.format(this)
}

fun Date.formatAsTime(): String {
    val formatter = SimpleDateFormat("h:mm a", Locale.US)
    return formatter.format(this)
}

// Money helpers for backend integration (amountMinor = cents)
fun Double.toAmountMinor(): Long {
    return (this * 100).toLong()
}

fun Long.fromAmountMinor(): Double {
    return this / 100.0
}

fun Long.formatAmountMinorAsCurrency(): String {
    return this.fromAmountMinor().formatAsCurrency()
}

File: ios/ExpenseTracker/Data/DemoData.swift
import Foundation

// MARK: - Data Models

struct Transaction: Identifiable {
    let id = UUID()
    let title: String
    let amount: Double
    let isExpense: Bool
    let category: String
    let date: Date
    let familyMember: String
}

struct Category: Identifiable {
    let id = UUID()
    let name: String
    let icon: String
    let color: String
    let isExpense: Bool
}

struct SavingsGoal: Identifiable {
    let id = UUID()
    let name: String
    let currentAmount: Double
    let targetAmount: Double
    
    var progress: Double {
        return currentAmount / targetAmount
    }
    
    var progressPercent: String {
        return "\(Int(progress * 100))%"
    }
}

struct FamilyMember: Identifiable {
    let id = UUID()
    let name: String
    let role: String
    let avatar: String
}

struct BudgetCategory: Identifiable {
    let id = UUID()
    let name: String
    let limit: Double
    let spent: Double
    
    var progress: Double {
        return spent / limit
    }
}

struct RecurringExpense: Identifiable {
    let id = UUID()
    let name: String
    let amount: Double
    let category: String
    let frequency: String
    let nextDate: Date
}

// MARK: - Backend DTO Models

struct LedgerCreateDTO: Codable {
    let type: String // "EXPENSE" or "INCOME"
    let amountMinor: Int64
    let currency: String
    let categoryId: String
    let memberId: String
    let occurredAt: String // ISO 8601 timestamp
    let notes: String?
}

struct RecurringRuleDTO: Codable {
    let id: String?
    let familyId: String
    let name: String
    let type: String // "EXPENSE" or "INCOME"
    let amountMinor: Int64
    let currency: String
    let categoryId: String
    let memberId: String
    let frequency: FrequencyDTO
    let startDate: String // ISO date
    let endDate: String?
    let timezone: String
    let nextRunAt: String?
    let isPaused: Bool
}

struct FrequencyDTO: Codable {
    let unit: String // "WEEKLY", "MONTHLY", "YEARLY"
    let interval: Int
    let byMonthDay: [Int]?
}

struct BudgetDTO: Codable {
    let id: String?
    let name: String
    let period: PeriodDTO
    let overallLimitMinor: Int64
    let includeRecurring: Bool
    let alertThresholdPct: Int
    let perCategory: [CategoryBudgetDTO]
}

struct PeriodDTO: Codable {
    let type: String // "MONTH", "QUARTER", "YEAR", "CUSTOM"
    let start: String // ISO date
    let end: String // ISO date
}

struct CategoryBudgetDTO: Codable {
    let categoryId: String
    let limitMinor: Int64
}

struct BudgetSpendDTO: Codable {
    let period: PeriodDTO
    let overall: SpendingSummaryDTO
    let byCategory: [CategorySpendingDTO]
}

struct SpendingSummaryDTO: Codable {
    let limitMinor: Int64
    let spentMinor: Int64
}

struct CategorySpendingDTO: Codable {
    let categoryId: String
    let limitMinor: Int64
    let spentMinor: Int64
}

struct ReportSummaryDTO: Codable {
    let totalIncomeMinor: Int64
    let totalExpensesMinor: Int64
    let balanceMinor: Int64
    let perCategory: [CategorySummaryDTO]
    let perMonth: [MonthlySummaryDTO]
    let recentTransactions: [TransactionSummaryDTO]
}

struct CategorySummaryDTO: Codable {
    let categoryId: String
    let spentMinor: Int64
}

struct MonthlySummaryDTO: Codable {
    let month: String // "2025-10"
    let incomeMinor: Int64
    let expenseMinor: Int64
}

struct TransactionSummaryDTO: Codable {
    let id: String
    let categoryId: String
    let amountMinor: Int64
}

struct ExportResponse: Codable {
    let exportId: String
}

// MARK: - Demo Data

class DemoData {
    static let shared = DemoData()
    
    private init() {}
    
    let totalIncome: Double = 12500.00
    let totalExpenses: Double = 8200.00
    
    var balance: Double {
        return totalIncome - totalExpenses
    }
    
    let transactions = [
        Transaction(title: "Salary", amount: 5000.00, isExpense: false, category: "Income", date: Date(), familyMember: "John"),
        Transaction(title: "Grocery Shopping", amount: 156.78, isExpense: true, category: "Groceries", date: Date(), familyMember: "Sarah"),
        Transaction(title: "Gas Station", amount: 45.50, isExpense: true, category: "Transportation", date: Date(), familyMember: "John"),
        Transaction(title: "Restaurant", amount: 89.32, isExpense: true, category: "Dining", date: Date(), familyMember: "Sarah"),
        Transaction(title: "Freelance Work", amount: 800.00, isExpense: false, category: "Income", date: Date(), familyMember: "John"),
    ]
    
    let categories = [
        Category(name: "Groceries", icon: "cart", color: "accentSuccess", isExpense: true),
        Category(name: "Transportation", icon: "car", color: "accentWarning", isExpense: true),
        Category(name: "Dining", icon: "fork.knife", color: "accentInfo", isExpense: true),
        Category(name: "Entertainment", icon: "tv", color: "primary", isExpense: true),
        Category(name: "Utilities", icon: "house", color: "expense", isExpense: true),
        Category(name: "Salary", icon: "dollarsign.circle", color: "income", isExpense: false),
        Category(name: "Freelance", icon: "briefcase", color: "income", isExpense: false),
    ]
    
    let savingsGoals = [
        SavingsGoal(name: "Emergency Fund", currentAmount: 2500.00, targetAmount: 5000.00),
        SavingsGoal(name: "Vacation", currentAmount: 1200.00, targetAmount: 3000.00),
        SavingsGoal(name: "New Car", currentAmount: 4500.00, targetAmount: 6000.00),
        SavingsGoal(name: "Home Renovation", currentAmount: 900.00, targetAmount: 3000.00),
    ]
    
    let familyMembers = [
        FamilyMember(name: "John Smith", role: "Admin", avatar: "person.circle.fill"),
        FamilyMember(name: "Sarah Smith", role: "Member", avatar: "person.circle.fill"),
        FamilyMember(name: "Emma Smith", role: "Member", avatar: "person.circle.fill"),
    ]
    
    let budgetCategories = [
        BudgetCategory(name: "Groceries", limit: 500.00, spent: 325.00),
        BudgetCategory(name: "Transportation", limit: 300.00, spent: 245.00),
        BudgetCategory(name: "Dining", limit: 200.00, spent: 89.32),
        BudgetCategory(name: "Entertainment", limit: 150.00, spent: 67.50),
    ]
    
    let recurringExpenses = [
        RecurringExpense(name: "Netflix", amount: 15.99, category: "Entertainment", frequency: "Monthly", nextDate: Calendar.current.date(byAdding: .day, value: 5, to: Date()) ?? Date()),
        RecurringExpense(name: "Gym Membership", amount: 29.99, category: "Health", frequency: "Monthly", nextDate: Calendar.current.date(byAdding: .day, value: 12, to: Date()) ?? Date()),
        RecurringExpense(name: "Phone Bill", amount: 89.00, category: "Utilities", frequency: "Monthly", nextDate: Calendar.current.date(byAdding: .day, value: 8, to: Date()) ?? Date()),
    ]
}